internal/receiver/discoveryreceiver/evaluator.go

```diff
@@ -0,0 +1,173 @@
+// Copyright Splunk, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package discoveryreceiver
+
+import (
+	"encoding/base64"
+	"fmt"
+	"regexp"
+	"sync"
+
+	"github.com/antonmedv/expr"
+	"github.com/antonmedv/expr/vm"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/extension/observer"
+	"go.opentelemetry.io/collector/config"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.uber.org/zap"
+	"gopkg.in/yaml.v2"
+)
+
+var (
+	observerIDAttr = "discovery.observer.id"
+)
+
+// evaluator is the base status matcher that determines if telemetry warrants emitting a matching log record.
+// It also provides embedded config correlation that its embedding structs will utilize.
+type evaluator struct {
+	logger       *zap.Logger
+	config       *Config
+	correlations correlationStore
+	// if match.FirstOnly this ~sync.Map(map[string]struct{}) keeps track of
+	// whether we've already emitted a record for the statement and can skip processing.
+	alreadyLogged *sync.Map
+	exprEnv       func(pattern string) map[string]any
+}
+
+// evaluateMatch parses the provided Match and returns whether it warrants a status log record
+func (e *evaluator) evaluateMatch(match Match, pattern, status string, receiverID config.ComponentID, endpointID observer.EndpointID) (shouldLog bool, err error) {
+	var matchFunc func(p string) (bool, error)
+	var matchPattern string
+
+	switch {
+	case match.Strict != "":
+		matchPattern = match.Strict
+		matchFunc = func(p string) (bool, error) {
+			return p == match.Strict, nil
+		}
+	case match.Regexp != "":
+		matchPattern = match.Regexp
+		var re *regexp.Regexp
+		if re, err = regexp.Compile(matchPattern); err != nil {
+			err = fmt.Errorf("invalid match regexp statement: %w", err)
+		} else {
+			matchFunc = func(p string) (bool, error) { return re.MatchString(p), nil }
+		}
+	case match.Expr != "":
+		matchPattern = match.Expr
+		var program *vm.Program
+		// TODO: cache compiled programs for performance benefit
+		if program, err = expr.Compile(match.Expr, expr.Env(e.exprEnv(pattern))); err != nil {
+			err = fmt.Errorf("invalid match expr statement: %w", err)
+		} else {
+			matchFunc = func(p string) (bool, error) {
+				ret, runErr := vm.Run(program, e.exprEnv(p))
+				if runErr != nil {
+					return false, runErr
+				}
+				return ret.(bool), nil
+			}
+		}
+	default:
+		err = fmt.Errorf("no valid match field provided")
```

Comment:
I'm not seeing where err is defined?

Reply:
named result parameters: https://go.dev/tour/basics/7

I chose to name `shouldLog` to emphasize that it's not just a match, but whether it should result in a log record (because of first only).

Reply:
`Naked return statements should be used only in short functions` 😅  maybe I should refine a bit.

Reply:
updated